### PR TITLE
Refactor cached or synchronous completion.

### PR DESCRIPTION
**Pull request by tigersoldier**
For synchronous completion, `company-lsp` should return a list of candidates
immediately. The cons (:async . FUNC) form should only be returned when
the cache is not hit and asynchronous completion is enabled.

### CODE REVIEW

**Comment by tigersoldier** [comment]
@sebastiencs @dgutov are you interested in reviewing this PR? :)

File: company-lsp.el [hunk 1]
```diff
@@ -244,10 +244,12 @@ CANDIDATE is a string returned by `company-lsp--make-candidate'."
     (unless (eq (point) point-before-post-complete)
       (setq this-command 'self-insert-command))))
 
-(defun company-lsp--on-completion (response prefix callback)
-  "Give the server RESPONSE to company's CALLBACK.
+(defun company-lsp--on-completion (response prefix)
+  "Handle completion RESPONSE.
 
-PREFIX is a stirng of the prefix when the completion is requested."
+PREFIX is a stirng of the prefix when the completion is requested.
```
Comment (dgutov) [inline company-lsp.el]: Pre-existing typo: "stirng"

File: company-lsp.el [hunk 2]
```diff
@@ -305,22 +328,15 @@ See the documentation of `company-backends' for COMMAND and ARG."
...
-                     (setq company-lsp--line-backup
-                           (buffer-substring (line-beginning-position) (line-end-position)))
-                     (if company-lsp-async
-                         (lsp--send-request-async req
-                                                  (lambda (resp)
-                                                    (company-lsp--on-completion resp arg callback)))
-                       (company-lsp--on-completion (lsp--send-request req)
-                                                   arg
-                                                   callback))))))))
+     (setq company-lsp--line-backup
```
Comment (dgutov) [inline company-lsp.el]: I'm not sure why you need to save this, TBH.

**Comment by dgutov** [comment]
Looks okay to me.

File: company-lsp.el [hunk 3]
```diff
@@ -305,22 +328,15 @@ See the documentation of `company-backends' for COMMAND and ARG."
...
+           (buffer-substring (line-beginning-position) (line-end-position)))
+     (let ((cache (company-lsp--cache-get arg)))
+       (if cache
+           (plist-get cache :candidates)
+         (if company-lsp-async
+             (cons :async (lambda (callback)
+                            (company-lsp--candidates-async arg callback)))
+           (company-lsp--candidates-sync arg)))))
```
Comment (sebastiencs) [inline company-lsp.el]: Thanks for asking. Looks good.
I would just write:
```el
(or (plist-get (company-lsp--cache-get arg) :candidates)
    (and company-lsp-async
         (cons :async (lambda (callback)
                        (company-lsp--candidates-async arg callback))))
    (company-lsp--candidates-sync arg))

```

Comment (tigersoldier) [reply company-lsp.el]: Thanks. Done.

File: company-lsp.el (hunk 2 above)
Comment (tigersoldier) [reply company-lsp.el]: That's a sign for documentation. Added.

File: company-lsp.el (hunk 1 above)
Comment (tigersoldier) [reply company-lsp.el]: Done

File: company-lsp.el [hunk 4]
```diff
@@ -305,22 +328,18 @@ See the documentation of `company-backends' for COMMAND and ARG."
...
-                                                  (lambda (resp)
-                                                    (company-lsp--on-completion resp arg callback)))
-                       (company-lsp--on-completion (lsp--send-request req)
-                                                   arg
-                                                   callback))))))))
+     ;; If the completion items in the response have textEdit action populated,
+     ;; we'll apply them in `company-lsp--post-completion'. However, textEdit
+     ;; actions only apply to the pre-completion content. We backup the current
+     ;; line and restore it after company completion is done, so the content
+     ;; is restored and textEdit actions can be applied.
```
Comment (dgutov) [inline company-lsp.el]: Which pre-completion context is this referring to? No prefix typed at all? With prefix, but without its last character? Or either, depending on the situation?

Comment (tigersoldier) [reply company-lsp.el]: With prefix, but without the completed suffix. Essentially we want to undo the completion by company.

Comment (dgutov) [reply company-lsp.el]: Hmm, yeah, that's suboptimal. Not sure what to suggest. Company could provide more info to `post-completion`, though.

**Pull request merged**